tsdb/db.go

```diff
@@ -583,6 +583,29 @@ func open(dir string, l log.Logger, r prometheus.Registerer, opts *Options, rngs
 	if err := removeBestEffortTmpDirs(l, dir); err != nil {
 		return nil, errors.Wrap(err, "remove tmp dirs")
 	}
+	if opts.WALSegmentSize >= 0 {
+		// Truncate WAL on startup.
+		first, last, err := wal.Segments(walDir)
+		if err != nil {
+			return nil, errors.Wrap(err, "wal-truncate")
+		}
+		last--
+		if !(last < 0 || first+(last-first)*2/3 < first) {
+			w, err := wal.New(l, nil, walDir, false)
+			if err != nil {
+				return nil, errors.Wrap(err, "wal creation")
+			}
+			if _, err = wal.Checkpoint(l, w, first, last, nil, int64(math.MinInt64)); err != nil {
```

Comment:
A checkpoint can take a few minutes, we probably shouldn't delay startup for it.

Reply:
I guess, startup happens after init head. But before init head, reload is called which has truncation. That means that if this is a point of delay, then the delay was already happening, meaning this won't add to delays.

Reply:
Truncation is fast, checkpointing is not.

Reply:
Head truncate function does checkpointing of wal as well.

Reply:
So if the reload is already doing this 70 lines below, why do we need to duplicate that code here again?

Reply:
Afaik, So as to make truncation of wals happen at the earliest and get additional space in constrainted environments.

Reply:
We don't do any sample writes until this whole function is completed, so this doesn't make any practical difference.